content/post/ddos-detection.md

```diff
@@ -0,0 +1,413 @@
+
+![503][503 service unavailable]
+(source: https://andrewodendaal.com/503-service-unavailable-back-end-server-is-at-capacity/)
+![502 bad gateway][502 bad gateway]
+![this site cannot be reached][this site cannot be reached]
+
+Well that sucks.
+
+There are a lot of reasons why a service can become unavailable or unreachable, and many of them are benign, like hardware hiccups, or bad code, or maybe a messed up routing table; one reason, however, is not benign: a denial of service (DoS) attack, or more commonly these days, a distributed denial of service (DDoS) attack.
+
+According to [Cloudflare][cloudflare what is ddos],
+
+> [a] distributed denial-of-service (DDoS) attack is a malicious attempt to disrupt normal traffic of a targeted server, service or network by overwhelming the target or its surrounding infrastructure with a flood of Internet traffic.
+
+In other words, the attack utilizes the fact that responding to each _request_ takes up a fraction of your system's finite resources, so if an attacker can generate sufficiently many requests, they can use up _all_ of your resources, leaving none available to service legitimate requests from your users. But that seems easy to defend against, right? Just throttle traffic by user or session! Ah, but this is where the "distributed" part comes in. In order to avoid being throttled, an attacker will use a large number of (probably) compromised machines to send service requests. At that point, it can become difficult or expensive to determine which new request from an unknown user may be legitimate and which may be malicious.
+
+Responding to attacks is an art in itself, but the first prerequisite is knowing that you're under attack. Early detection is key.
+
+## Detecting an Attack is Underway
+
+In order to detect an attack, we need two things: a model for "good" or "normal" behaviour, and a way to measure our services' current behaviour to compare against the "normal" one. If the difference is too great, we can mark the present data point as anomalous, and in the DDoS attack case, we will mark the server as "under attack". This is more or less the essence of anomaly detection.
+Since DDoS attacks typically involve a sharp increase in traffic, we can use the current traffic to derive a "normal" model on the fly, using the ongoing metrics from our servers. Once a normal model is established, we can look for a sharp increase in the number of requests and number of clients that a server is handling as an indicator for a potential DDoS attack. To do this, we will compute a weighted mean and a weighted standard deviation over a the last 60 seconds of metrics data per server. And then for each new data point in the current second of data, we will try to predict if, once it is complete, its difference from the mean is greater than 2 standard deviations. If it is, we will mark the server as under attack.
```

Comment:
should be a newline before each paragraph for correct formatting. All your paragraphs are formatted as a giant blob of text in here. you need to check out the preview.